### ISSUE TITLE
Prevent race conditions and control how to process the sequence of events

### ISSUE THREAD

**Issue by sokolej79** [rrousselGit/river_pod#1215]
Riverpod should have a notifier class that would prevent race conditions and control how to process the sequence of events. Something like a bloc packag with bloc_concurrency.
This would allow you to control different scenarios of how to process events.
Event control is usually: cancell last when new event, allow to finished last then next, parallel and sequential.
Most state management packages allow event control: redux with redux_epics, blox with bloc_concurrency, rx with rx operators switchMap, concatMap, mergeMap, exhaustMap ...
If this doesn't make sense, is it possible that riverpod could use a bloc package for these scenarios without flutter_bloc only bloc_core and vice versa bloc class can get state from riverpod statenotifier, changeNotifier within bloc class.
Good explanation for scenarios:
https://verygood.ventures/blog/how-to-use-bloc-with-streams-and-concurrency

**Comment by rrousselGit** [comment]
Could you clarify exactly what problem you're  trying to solve?

This issue propose a solution but does not explain the problem.

**Comment by sokolej79** [comment]
Problems scenarios:

1.) StateNotifier with multiple async actions that manipulate state  sequentially:
-user triger actions a, b... that should change state sequntally as action triggered
-a could take longer to finish that b and b will update state beore a
-rxjs uses concatMap for that scenario
-bloc uses package bloc_concurrency eventTransformer sequential

2.) StateNotifier search - debounceTime and cancell previouslly request:
-user triggers search multiple times, previouslly request should be cancelled and optionally set debounce time on search
-rxjs uses distinctUntillChanges (dont execute same search), debouncetime (execute ater debounce time period) and switchmap that cancel previouslly requet on new search when for that scenario
-bloc uses package bloc_concurrency eventTransformer droppable

3.) Command button with action should wait for first action to finish if user clicks again and action did not finished than igore each next request until firts finishes
-rxjs uses exhaustMap for that scenario
-bloc uses package bloc_concurrency eventTransformer restartable - process only the latest event and cancel previous event handlers

How to achive this scenarios with riverpod SStateNotifier without behaviorsubject and rx?

4.) Use bloc and riverpod in same project:
If you have some widgets with bloc package and others with riverpod in same app how to connect, listen state in riverpod from bloc and  bloc from riverpod.
Is it possible?

**Comment by rrousselGit** [comment]
1.) StateNotifier with multiple async actions that manipulate state sequentially:


Depends on the case more specifically. It could be that you can refactor it such that the problem disappears,

> 2.) StateNotifier search - debounceTime and cancell previouslly request:

You likely wouldn't be using StateNotifier to begin with. The fetching would be done by FutureProvider, and debouncing/cancelling network requests too.

I made a tweet about this recently https://twitter.com/remi_rousselet/status/1493522569005305858

This snippet handles both:
- debouncing
- cancelling network request if the user leaves the screen while a request is pending
- cancelling network request  if the search query changes the screen while a request is pending

> 3.) Command button with action should wait for first action to finish if user clicks again and action did not finished than igore each next request until firts finishes


That's broad. There are numerous solutions, such as doing it within your button directly by implementing:

![image](https://user-images.githubusercontent.com/20165741/155314904-0788581a-c884-40b3-a0d6-347d93dcab91.png)


In which case, chances are this would be completely independent from riverpod (and be similar to RefreshIndicator)

**Comment by sokolej79** [comment]
Ok. Thaaß40

**Comment by sokolej79** [comment]
Ok, thanks for reply. Otherwise Riverpod is great package. Closing.

**Comment by BenjiFarquhar** [comment]
@rrousselGit  I've noticed there is a race condition with `Notifier` and `StateNotifier` in several places in my code. Sometimes calling `xNotifier.state = newState;` and then watching that state in another widget, the widget does not contain the `newState` yet on the very next build, like it is building too soon, but it is building after the line `xNotifier.state = newState;` has executed. I swapped out the `Notifier` out for a `flutter_bloc` `cubit`, and it resolved the issue so there is some kind of bug with `Notifier`.
